File: cpp/dialogs/file_browser.cpp
#include "file_browser.hpp"

#include "message_box.hpp"

#include <algorithm>
#include <chrono>
#include <cctype>
#include <regex>
#include <sstream>
#include <system_error>
#include <utility>
#include <vector>

namespace ui::dialogs {

namespace {

constexpr auto kFileBrowserDoubleClickInterval = std::chrono::milliseconds(400);

std::string displayName(const std::filesystem::directory_entry& entry) {
    const std::string name = entry.path().filename().string();
    return entry.is_directory() ? name + "/" : name;
}

std::string trim(std::string text) {
    const auto notSpace = [](unsigned char ch) {
        return !std::isspace(ch);
    };
    text.erase(text.begin(), std::find_if(text.begin(), text.end(), notSpace));
    text.erase(std::find_if(text.rbegin(), text.rend(), notSpace).base(), text.end());
    return text;
}

}  // namespace

FileBrowser::FileBrowser(std::filesystem::path rootDir, std::filesystem::path startPath, std::string mode,
                         FileBrowserCallback onClose, std::string initialFilename, std::string extensionFilter)
    : rootDir_(std::filesystem::weakly_canonical(std::move(rootDir))),
      mode_(std::move(mode)),
      onClose_(std::move(onClose)),
      extensionFilter_(trim(std::move(extensionFilter))) {
    std::filesystem::path requestedStart = startPath.empty() ? rootDir_ : std::move(startPath);
    if (std::filesystem::is_regular_file(requestedStart)) {
        currentDir_ = clampDir(requestedStart.parent_path());
        if (initialFilename.empty()) {
            initialFilename = requestedStart.filename().string();
        }
    } else {
        currentDir_ = clampDir(std::move(requestedStart));
    }
    initialFilename_ = std::move(initialFilename);
    dialog_header->setText(mode_ == "save" ? "Save As" : "Open File");
    open_btn->setTitle(mode_ == "save" ? "Save" : "Open");
    if (mode_ == "save") {
        filename->setValue(initialFilename_);
        moveFilenameCursorToEnd();
    }
    refreshEntries();
}

void FileBrowser::refreshEntries() {
    std::vector<std::string> rows{".."};
    std::vector<std::string> disabledRows;

    std::vector<std::filesystem::directory_entry> dirs;
    std::vector<std::filesystem::directory_entry> files;
    std::error_code error;
    for (const auto& entry : std::filesystem::directory_iterator(currentDir_, error)) {
        if (entry.is_directory()) {
            dirs.push_back(entry);
        } else if (entry.is_regular_file()) {
            files.push_back(entry);
        }
    }
    const auto byName = [](const auto& lhs, const auto& rhs) {
        return lhs.path().filename().string() < rhs.path().filename().string();
    };
    std::sort(dirs.begin(), dirs.end(), byName);
    std::sort(files.begin(), files.end(), byName);
    for (const auto& dir : dirs) {
        rows.push_back(displayName(dir));
    }
    for (const auto& file : files) {
        const std::string name = displayName(file);
        rows.push_back(name);
        if (!pathMatchesFilter(file.path())) {
            disabledRows.push_back(name);
        }
    }

    path_label->setText(currentDir_.string());
    entries->setOptions(std::move(rows));
    entries->setDisabledValues(std::move(disabledRows));
    const auto& options = entries->options();
    auto selected = std::find(options.begin(), options.end(), initialFilename_);
    entries->setSelectedIndex(selected == options.end() ? 0 : static_cast<int>(selected - options.begin()));
    lastClickedEntry_ = -1;
    previewSelected();
}

bool FileBrowser::acceptCurrent() {
    const std::filesystem::path path = selectedPath();
    if (path.empty()) {
        return false;
    }
    if (std::filesystem::is_directory(path)) {
        currentDir_ = clampDir(path);
        refreshEntries();
        return true;
    }
    const std::string selectedName = path.filename().string();
    const std::string name = trim(filename->value().empty() ? selectedName : filename->value());
    if (name.empty()) {
        return false;
    }
    const std::filesystem::path namedPath = currentDir_ / name;
    if (mode_ == "open" && std::filesystem::is_regular_file(namedPath) && clampDir(namedPath.parent_path()) == currentDir_) {
        if (!pathMatchesFilter(namedPath)) {
            return false;
        }
        close(namedPath.string());
        return true;
    }
    if (mode_ == "save") {
        return acceptFilename();
    }
    return false;
}

bool FileBrowser::acceptFilename() {
    const std::string name = trim(filename->value());
    if (name.empty()) {
        return false;
    }
    const std::filesystem::path path = currentDir_ / name;
    if (std::filesystem::is_directory(path) || clampDir(path.parent_path()) != currentDir_) {
        return false;
    }
    if (std::filesystem::is_regular_file(path)) {
        confirmOverwrite(path);
        return true;
    }
    close(path.string());
    return true;
}

void FileBrowser::selectEntry(int index) {
    entries->setSelectedIndex(index);
    previewSelected();
}

bool FileBrowser::selectedEntryIsDirectory() const {
    return entryIndexIsDirectory(entries->selectedIndex());
}

bool FileBrowser::entryIndexIsDirectory(int index) const {
    const auto& options = entries->options();
    if (index < 0 || index >= static_cast<int>(options.size())) {
        return false;
    }
    const std::string& selected = options[static_cast<std::size_t>(index)];
    return selected == ".." || selected.ends_with("/");
}

bool FileBrowser::handleEntryMousePress(Point point) {
    const Rect frame = entries->frame();
    if (!frame.contains(point)) {
        return false;
    }
    const int index = entries->scrollOffset() + point.row - frame.row;
    const auto now = std::chrono::steady_clock::now();
    const bool doubleClick = index == lastClickedEntry_ && now - lastEntryClickTime_ <= kFileBrowserDoubleClickInterval;
    selectEntry(index);
    lastClickedEntry_ = index;
    lastEntryClickTime_ = now;
    if (doubleClick && (entryIndexIsDirectory(index) || mode_ == "open")) {
        const bool accepted = acceptCurrent();
        lastClickedEntry_ = -1;
        return accepted;
    }
    return false;
}

void FileBrowser::moveFilenameCursorToEnd() {
    const int end = static_cast<int>(filename->value().size());
    filename->selectRange(end, end);
}

void FileBrowser::close(std::string path) {
    result_ = std::move(path);
    closed_ = true;
    if (onClose_) {
        onClose_(result_);
    }
}

int FileBrowser::run() {
    return runGeneratedWindow(*this, runtimeOptions());
}

GeneratedWindowFrameOptions FileBrowser::stackFrameOptions() {
    GeneratedWindowRuntimeOptions runtime = runtimeOptions();
    GeneratedWindowFrameOptions frame;
    frame.className = "FileBrowser";
    frame.initialFocusName = std::move(runtime.initialFocusName);
    frame.startInEditMode = runtime.startInEditMode;
    frame.onButton = std::move(runtime.onButton);
    frame.onKeyBeforeFocusedElement = std::move(runtime.onKeyBeforeFocusedElement);
    frame.onKeyBeforeFocused = std::move(runtime.onKeyBeforeFocused);
    frame.onKey = std::move(runtime.onKey);
    frame.onMousePressBeforeFocused = std::move(runtime.onMousePressBeforeFocused);
    frame.onTextChanged = std::move(runtime.onTextChanged);
    frame.onTextConfirmed = std::move(runtime.onTextConfirmed);
    frame.onSelectionChanged = std::move(runtime.onSelectionChanged);
    frame.onEditStarted = std::move(runtime.onEditStarted);
    frame.onFocusChanged = std::move(runtime.onFocusChanged);
    frame.shouldClose = std::move(runtime.shouldClose);
    return frame;
}

std::filesystem::path FileBrowser::clampDir(std::filesystem::path path) const {
    std::error_code error;
    std::filesystem::path canonical = std::filesystem::weakly_canonical(path, error);
    if (error || canonical.empty()) {
        canonical = rootDir_;
    }
    const std::string root = rootDir_.string();
    const std::string candidate = canonical.string();
    if (candidate.rfind(root, 0) != 0) {
        return rootDir_;
    }
    return canonical;
}

void FileBrowser::confirmOverwrite(const std::filesystem::path& path) {
    MessageBoxYesNo dialog(
        "Overwrite File",
        "File already exists:\n" + path.string() + "\n\nOverwrite it?");
    (void)dialog.run();
    if (dialog.result() == "yes") {
        close(path.string());
    }
}

void FileBrowser::previewSelected() {
    const auto& options = entries->options();
    if (options.empty()) {
        return;
    }
    const std::string& selected = options[static_cast<std::size_t>(entries->selectedIndex())];
    if (selected != "..") {
        filename->setValue(selected.ends_with("/") ? selected.substr(0, selected.size() - 1) : selected);
    } else if (mode_ == "open") {
        filename->setValue("");
    }
    moveFilenameCursorToEnd();
    updateOpenEnabled();
}

std::filesystem::path FileBrowser::selectedPath() const {
    const auto& options = entries->options();
    if (options.empty()) {
        return {};
    }
    const std::string& selected = options[static_cast<std::size_t>(entries->selectedIndex())];
    if (selected == "..") {
        return currentDir_.parent_path();
    }
    if (selected.ends_with("/")) {
        return currentDir_ / selected.substr(0, selected.size() - 1);
    }
    return currentDir_ / selected;
}

bool FileBrowser::pathMatchesFilter(const std::filesystem::path& path) const {
    if (extensionFilter_.empty()) {
        return true;
    }
    std::string ext = path.extension().string();
    if (!ext.empty() && ext.front() == '.') {
        ext.erase(ext.begin());
    }
    if (ext.empty()) {
        return false;
    }
    try {
        return std::regex_match(ext, std::regex(extensionFilter_, std::regex_constants::icase));
    } catch (const std::regex_error&) {
        std::istringstream stream(extensionFilter_);
        std::string part;
        while (std::getline(stream, part, '|')) {
            part = trim(std::move(part));
            if (!part.empty() && part.front() == '.') {
                part.erase(part.begin());
            }
            std::transform(part.begin(), part.end(), part.begin(), [](unsigned char ch) {
                return static_cast<char>(std::tolower(ch));
            });
            std::string lowered = ext;
            std::transform(lowered.begin(), lowered.end(), lowered.begin(), [](unsigned char ch) {
                return static_cast<char>(std::tolower(ch));
            });
            if (lowered == part) {
                return true;
            }
        }
    }
    return false;
}

std::filesystem::path FileBrowser::selectedPathForOpenState() const {
    const std::string name = trim(filename->value());
    if (!name.empty()) {
        return currentDir_ / name;
    }
    const auto& options = entries->options();
    if (options.empty()) {
        return {};
    }
    const std::string& selected = options[static_cast<std::size_t>(entries->selectedIndex())];
    if (selected.empty()) {
        return {};
    }
    if (selected == "..") {
        return clampDir(currentDir_.parent_path());
    }
    if (selected.ends_with("/")) {
        return currentDir_ / selected.substr(0, selected.size() - 1);
    }
    return currentDir_ / selected;
}

void FileBrowser::updateOpenEnabled() {
    if (mode_ == "save") {
        open_btn->setEnabled(!trim(filename->value()).empty());
        return;
    }
    const std::filesystem::path path = selectedPathForOpenState();
    if (path.empty()) {
        open_btn->setEnabled(false);
        return;
    }
    if (std::filesystem::is_directory(path)) {
        open_btn->setEnabled(true);
        return;
    }
    open_btn->setEnabled(std::filesystem::is_regular_file(path) && pathMatchesFilter(path));
}

GeneratedWindowRuntimeOptions FileBrowser::runtimeOptions() {
    GeneratedWindowRuntimeOptions options;
    options.initialFocusName = "entries";
    options.startInEditMode = true;
    options.onButton = [this](std::string_view name) {
        if (name == "open_btn") {
            if (mode_ == "save") {
                (void)acceptFilename();
            } else {
                (void)acceptCurrent();
            }
        } else if (name == "close_btn") {
            close("");
        }
    };
    options.onTextChanged = [this](std::string_view name, std::string_view) {
        if (name == "entries") {
            previewSelected();
        } else if (name == "filename") {
            updateOpenEnabled();
        }
    };
    options.onSelectionChanged = [this](std::string_view name, const std::vector<std::string>&) {
        if (name == "entries") {
            previewSelected();
        }
    };
    options.onTextConfirmed = [this](std::string_view name, std::string_view) {
        if (name == "entries") {
            previewSelected();
        }
    };
    options.onKeyBeforeFocusedElement = [this](std::string_view key, std::string_view name, bool editMode) {
        if (name == "entries" && editMode && key == "Enter") {
            if (selectedEntryIsDirectory()) {
                (void)acceptCurrent();
            }
            return true;
        }
        return false;
    };
    options.onEditStarted = [this](std::string_view name) {
        if (name == "filename") {
            moveFilenameCursorToEnd();
        }
    };
    options.onMousePressBeforeFocused = [this](Point point) {
        return handleEntryMousePress(point);
    };
    options.onKey = [this](std::string_view key) {
        if (key == "Escape") {
            close("");
            return true;
        }
        return false;
    };
    options.shouldClose = [this] {
        return closed_;
    };
    return options;
}

}  // namespace ui::dialogs

File: java/src/main/java/uimd/FileBrowser.java
package uimd;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.function.Consumer;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;
import java.util.stream.Stream;

public final class FileBrowser extends FileBrowserUI
{
    public static final String MODE_OPEN = "open";
    public static final String MODE_SAVE = "save";

    private static final String PARENT_ENTRY = "..";

    private final Path rootDir;
    private final String browserMode;
    private final Consumer<Path> onClose;
    private final String extensionFilter;
    private final String initialFilename;
    private Path currentDir;
    private Path result;
    private boolean closed;
    private GeneratedWindowStack activeStack;

    public FileBrowser(Path rootDir)
    {
        this(rootDir, null, MODE_OPEN, null, "", "");
    }

    public FileBrowser(
        Path rootDir,
        Path startPath,
        String mode,
        Consumer<Path> onClose,
        String initialFilename,
        String extensionFilter)
    {
        if (rootDir == null)
        {
            throw new IllegalArgumentException("rootDir must not be null");
        }
        this.rootDir = rootDir.toAbsolutePath().normalize();
        browserMode = MODE_SAVE.equals(mode) ? MODE_SAVE : MODE_OPEN;
        this.onClose = onClose;
        this.extensionFilter = extensionFilter == null ? "" : extensionFilter.trim();

        Path requestedStart = (startPath == null ? this.rootDir : startPath)
            .toAbsolutePath()
            .normalize();
        String requestedFilename = initialFilename == null ? "" : initialFilename;
        if (Files.isRegularFile(requestedStart))
        {
            currentDir = clampDir(requestedStart.getParent());
            if (requestedFilename.isEmpty() && requestedStart.getFileName() != null)
            {
                requestedFilename = requestedStart.getFileName().toString();
            }
        }
        else
        {
            currentDir = clampDir(requestedStart);
        }
        this.initialFilename = requestedFilename;
        dialog_header.setText(MODE_SAVE.equals(browserMode) ? "Save As" : "Open File");
        open_btn.setTitle(MODE_SAVE.equals(browserMode) ? "Save" : "Open");
        if (MODE_SAVE.equals(browserMode))
        {
            filename.setValue(this.initialFilename);
            moveFilenameCursorToEnd();
        }
        refreshEntries();
    }

    public Path currentDir()
    {
        return currentDir;
    }

    public Path result()
    {
        return result;
    }

    public boolean closed()
    {
        return closed;
    }

    public int run(String[] arguments)
    {
        GeneratedWindowStack stack = new GeneratedWindowStack();
        GeneratedWindowRuntimeOptions options = runtimeOptions(stack);
        return GeneratedWindowRuntime.runGeneratedWindow(this, options, arguments);
    }

    public GeneratedWindowRuntimeOptions runtimeOptions(GeneratedWindowStack stack)
    {
        activeStack = stack;
        GeneratedWindowRuntimeOptions options = runtimeOptions();
        options.setWindowStack(stack);
        return options;
    }

    @Override
    public GeneratedWindowRuntimeOptions runtimeOptions()
    {
        GeneratedWindowRuntimeOptions options = super.runtimeOptions();
        options.setInitialFocusName("entries");
        options.setStartInEditMode(true);
        options.setOnMousePressBeforeFocused(this::handleEntryMousePress);
        options.setOnEditStarted(name ->
        {
            if ("filename".equals(name))
            {
                moveFilenameCursorToEnd();
            }
        });
        options.setShouldClose(this::closed);
        return options;
    }

    @Override
    public boolean handleGeneratedListBoxItemActivate(
        String name,
        String elementId,
        int index,
        String value)
    {
        if (!"entries".equals(name))
        {
            return false;
        }
        entries.setSelectedIndex(index);
        entries.hideActiveItem();
        previewSelected();
        return selectedEntryIsDirectory() && acceptCurrent();
    }

    @Override
    public boolean onPreviewKey(KeyEvent event)
    {
        if (!"Escape".equals(event.key()) || event.editMode())
        {
            return false;
        }
        close(null);
        return true;
    }

    public void refreshEntries()
    {
        List<Path> directories = new ArrayList<>();
        List<Path> files = new ArrayList<>();
        try (Stream<Path> stream = Files.list(currentDir))
        {
            stream.forEach(path ->
            {
                if (Files.isDirectory(path))
                {
                    directories.add(path);
                }
                else if (Files.isRegularFile(path))
                {
                    files.add(path);
                }
            });
        }
        catch (IOException exception)
        {
            directories.clear();
            files.clear();
        }
        Comparator<Path> byName = Comparator.comparing(
            path -> path.getFileName().toString());
        directories.sort(byName);
        files.sort(byName);

        List<String> rows = new ArrayList<>();
        List<String> disabledRows = new ArrayList<>();
        rows.add(PARENT_ENTRY);
        for (Path directory : directories)
        {
            rows.add(directory.getFileName() + "/");
        }
        for (Path file : files)
        {
            String name = file.getFileName().toString();
            rows.add(name);
            if (!pathMatchesFilter(file))
            {
                disabledRows.add(name);
            }
        }

        path_label.setText(currentDir.toString());
        entries.setOptions(rows);
        entries.setDisabledValues(disabledRows);
        int selectedIndex = rows.indexOf(initialFilename);
        entries.setSelectedIndex(selectedIndex >= 0 ? selectedIndex : 0);
        previewSelected();
    }

    public boolean acceptCurrent()
    {
        String selected = selectedEntry();
        String name = filename.value().trim();
        if (!name.isEmpty())
        {
            Path path = currentDir.resolve(name).toAbsolutePath().normalize();
            if (!path.startsWith(rootDir))
            {
                return false;
            }
            if (Files.isDirectory(path))
            {
                currentDir = clampDir(path);
                refreshEntries();
                return true;
            }
            if (MODE_OPEN.equals(browserMode)
                && (!Files.isRegularFile(path) || !pathMatchesFilter(path)))
            {
                return false;
            }
            close(path);
            return true;
        }
        if (PARENT_ENTRY.equals(selected))
        {
            currentDir = clampDir(currentDir.getParent());
            refreshEntries();
            return true;
        }
        if (selected.endsWith("/"))
        {
            currentDir = clampDir(currentDir.resolve(selected.substring(0, selected.length() - 1)));
            refreshEntries();
            return true;
        }
        if (selected.trim().isEmpty())
        {
            return false;
        }
        Path path = currentDir.resolve(selected.trim()).toAbsolutePath().normalize();
        if (!path.startsWith(rootDir)
            || (MODE_OPEN.equals(browserMode)
                && (!Files.isRegularFile(path) || !pathMatchesFilter(path))))
        {
            return false;
        }
        close(path);
        return true;
    }

    public boolean acceptFilename()
    {
        String name = filename.value().trim();
        if (name.isEmpty())
        {
            return false;
        }
        Path path = currentDir.resolve(name).toAbsolutePath().normalize();
        if (!path.startsWith(rootDir) || Files.isDirectory(path))
        {
            return false;
        }
        if (Files.isRegularFile(path))
        {
            confirmOverwrite(path);
            return true;
        }
        close(path);
        return true;
    }

    public void selectEntry(int index)
    {
        entries.setSelectedIndex(index);
        previewSelected();
    }

    public boolean selectedEntryIsDirectory()
    {
        return entryIndexIsDirectory(entries.selectedIndex());
    }

    public boolean entryIndexIsDirectory(int index)
    {
        if (index < 0 || index >= entries.options().size())
        {
            return false;
        }
        String selected = entries.options().get(index);
        return PARENT_ENTRY.equals(selected) || selected.endsWith("/");
    }

    public boolean handleEntryMousePress(Point point)
    {
        Rect frame = entries.frame();
        if (!frame.contains(point))
        {
            return false;
        }
        int index = entries.scrollOffset() + point.row() - frame.row();
        if (index < 0 || index >= entries.options().size())
        {
            return false;
        }
        selectEntry(index);
        return entryIndexIsDirectory(index) && acceptCurrent();
    }

    public void moveFilenameCursorToEnd()
    {
        int end = filename.value().length();
        filename.selectRange(end, end);
    }

    public void close(Path path)
    {
        result = path;
        closed = true;
        if (onClose != null)
        {
            onClose.accept(result);
        }
        GeneratedWindowStack stack = runtimeWindowStack();
        if (stack != null)
        {
            stack.remove(this);
        }
    }

    @Override
    protected void onEntriesSelectionChange(List<String> value)
    {
        previewSelected();
    }

    @Override
    protected void onFilenameChange(String value)
    {
        updateOpenEnabled();
    }

    @Override
    protected void onFilenameSubmit(String value)
    {
        if (MODE_SAVE.equals(browserMode))
        {
            acceptFilename();
        }
        else
        {
            acceptCurrent();
        }
    }

    @Override
    protected void onOpenBtnClick()
    {
        if (MODE_SAVE.equals(browserMode))
        {
            acceptFilename();
        }
        else
        {
            acceptCurrent();
        }
    }

    @Override
    protected void onCloseBtnClick()
    {
        close(null);
    }

    private Path clampDir(Path path)
    {
        if (path == null)
        {
            return rootDir;
        }
        Path candidate = path.toAbsolutePath().normalize();
        return candidate.startsWith(rootDir) && Files.isDirectory(candidate)
            ? candidate
            : rootDir;
    }

    private void confirmOverwrite(Path path)
    {
        MessageBoxYesNo dialog = new MessageBoxYesNo(
            "Overwrite File",
            "File already exists:\n" + path + "\n\nOverwrite it?",
            value ->
            {
                if ("yes".equals(value))
                {
                    close(path);
                }
            });
        GeneratedWindowStack stack = activeStack == null ? runtimeWindowStack() : activeStack;
        if (stack == null)
        {
            dialog.run(new String[0]);
            return;
        }
        GeneratedWindowRuntimeOptions options = dialog.runtimeOptions();
        options.setWindowStack(stack);
        stack.push(dialog, options);
    }

    private void previewSelected()
    {
        String selected = selectedEntry();
        if (!selected.isEmpty() && !PARENT_ENTRY.equals(selected))
        {
            filename.setValue(selected.endsWith("/")
                ? selected.substring(0, selected.length() - 1)
                : selected);
        }
        else if (MODE_OPEN.equals(browserMode))
        {
            filename.setValue("");
        }
        moveFilenameCursorToEnd();
        updateOpenEnabled();
    }

    private String selectedEntry()
    {
        List<String> selected = entries.selectedValues();
        return selected.isEmpty() ? "" : selected.get(0);
    }

    private Path selectedPathForOpenState()
    {
        String name = filename.value().trim();
        if (!name.isEmpty())
        {
            return currentDir.resolve(name).toAbsolutePath().normalize();
        }
        String selected = selectedEntry();
        if (selected.isEmpty())
        {
            return null;
        }
        if (PARENT_ENTRY.equals(selected))
        {
            return clampDir(currentDir.getParent());
        }
        String pathName = selected.endsWith("/")
            ? selected.substring(0, selected.length() - 1)
            : selected;
        return currentDir.resolve(pathName).toAbsolutePath().normalize();
    }

    private boolean pathMatchesFilter(Path path)
    {
        if (extensionFilter.isEmpty())
        {
            return true;
        }
        String name = path.getFileName() == null ? "" : path.getFileName().toString();
        int separator = name.lastIndexOf('.');
        String extension = separator < 0 || separator + 1 >= name.length()
            ? ""
            : name.substring(separator + 1).toLowerCase(Locale.ROOT);
        if (extension.isEmpty())
        {
            return false;
        }
        try
        {
            return Pattern.compile(extensionFilter, Pattern.CASE_INSENSITIVE)
                .matcher(extension)
                .matches();
        }
        catch (PatternSyntaxException exception)
        {
            for (String part : extensionFilter.split("\\|"))
            {
                String allowed = part.trim().toLowerCase(Locale.ROOT);
                if (allowed.startsWith("."))
                {
                    allowed = allowed.substring(1);
                }
                if (extension.equals(allowed))
                {
                    return true;
                }
            }
            return false;
        }
    }

    private void updateOpenEnabled()
    {
        if (MODE_SAVE.equals(browserMode))
        {
            open_btn.setEnabled(!filename.value().trim().isEmpty());
            return;
        }
        Path path = selectedPathForOpenState();
        if (path == null || !path.startsWith(rootDir))
        {
            open_btn.setEnabled(false);
        }
        else if (Files.isDirectory(path))
        {
            open_btn.setEnabled(true);
        }
        else
        {
            open_btn.setEnabled(Files.isRegularFile(path) && pathMatchesFilter(path));
        }
    }
}
